### PR TITLE
Error handling cleanup

### PR DESCRIPTION
**Pull request by slawlor**
A decent-sized rewrite of the error constructs to clarify a lot of the error patterns (i.e. storage throws clearer errors rather than just get/set). Additionally

1. Small utility functions which cannot throw errors should not return `Result<_, _>` and have been rewritten to not be error potential.
2. Digest serialization changed to `[u8; 32]` from `Vec<u8>` as we know the size and it makes it clearer to other developers downstream

### CODE REVIEW

File: akd/Cargo.toml [hunk 1]
```diff
@@ -32,18 +34,19 @@ bincode = { version = "1", optional = true }
 serde = { version = "1", features = ["derive"], optional = true }
 rand = { version = "0.7", optional = true }
 curve25519-dalek = { version = "3.2", optional = true }
-ed25519-dalek = { version = "1", optional = true}
-
-[target.'cfg(all(features = "serde", features = "vrf"))'.dependencies]
```
Comment (slawlor) [inline akd/Cargo.toml:37]: This is apparently not supported (and had typos anyways). Need to have builder utilize features on the crate to support it.

File: akd/Cargo.toml [hunk 2]
```diff
@@ -32,18 +34,19 @@ bincode = { version = "1", optional = true }
...
 
 [dev-dependencies]
 criterion = "0.3"
 serial_test = "0.5"
 proptest = "1.0"
 proptest-derive = "0.3"
+colored = { version = "2" }
+once_cell = { version = "1" }
+
+akd = { path =".", features = ["vrf", "public-tests"] }
```
Comment (slawlor) [inline akd/Cargo.toml:49]: We add ourself back in in the `test` profile in order to adjust the features on the create (i.e. enable the test-logic needed)

File: akd/src/errors.rs [hunk 3]
```diff
@@ -158,57 +169,47 @@ impl fmt::Display for AzksError {
...
     VerifyLookupProof(String),
     /// Key-History proof did not verify
     VerifyKeyHistoryProof(String),
     /// Tried to audit an invalid epoch range
     InvalidEpoch(String),
-    /// Error propagation
-    Storage(StorageError),
-    /// Error propagation for errors from VRF storage
-    VRFStorageErr(VRFStorageError),
+    /// AZKS not found in read-only directory mode
```
Comment (afterdusk) [inline akd/src/errors.rs:188]: Technically this error is thrown when we try to perform a write operation in read-only mode, right? This includes trying to publish. I think the error message can be clearer, especially in terms of what the bool stands for.

File: akd/src/errors.rs [hunk 4]
```diff
@@ -217,36 +218,61 @@ impl fmt::Display for DirectoryError {
...
+                write!(f, "Other storage error: {}", inner)
+            }
+        }
+    }
 }
 
 /// Represents a VRF-storage-layer error
 #[derive(PartialEq, Debug)]
-pub enum VRFStorageError {
+pub enum VrfError {
```
Comment (afterdusk) [inline akd/src/errors.rs:254]: I think the comments for the variants should be updated too!

File: akd/src/errors.rs [hunk 5]
```diff
@@ -217,36 +218,61 @@ impl fmt::Display for DirectoryError {
...
             }
-            Self::GetPK(error_string) => {
-                write!(f, "{}", error_string)
+            Self::PublicKey(error_string) => {
+                write!(f, "VRF public key: {}", error_string)
             }
-            Self::VRFErr(error_string) => {
-                write!(f, "{}", error_string)
+            Self::Verification(error_string) => {
+                write!(f, "VRF prooving or verifying: {}", error_string)
```
Comment (afterdusk) [inline akd/src/errors.rs:275]: nit: sp!

File: akd/src/directory.rs [hunk 6]
```diff
@@ -218,9 +212,7 @@ impl<S: Storage + Sync + Send, V: VRFKeyStorage> Directory<S, V> {
             if let Err(err) = self.storage.commit_transaction().await {
                 // ignore any rollback error(s)
                 let _ = self.storage.rollback_transaction().await;
-                return Err(AkdError::HistoryTreeNode(HistoryTreeNodeError::Storage(
-                    err,
-                )));
+                return Err(AkdError::Storage(err));
```
Comment (eozturk1) [inline akd/src/directory.rs:215]: nit: Maybe add the origin to the message? "Couldn't commit transaction in publish"

File: akd/src/directory.rs [hunk 7]
```diff
@@ -419,10 +408,10 @@ impl<S: Storage + Sync + Send, V: VRFKeyStorage> Directory<S, V> {
             }
             Ok(HistoryProof { proofs })
         } else {
-            Err(AkdError::Directory(DirectoryError::NonExistentUser(
-                username,
-                current_epoch,
-            )))
+            Err(AkdError::Storage(StorageError::NotFound(format!(
+                "User {} at epoch {}",
```
Comment (eozturk1) [inline akd/src/directory.rs:412]: Nice!

File: akd/src/errors.rs [hunk 8]
```diff
@@ -11,22 +11,22 @@ use core::fmt;
 use crate::node_state::NodeLabel;
 
 /// Symbolizes a AkdError, thrown by the akd.
-#[derive(Debug)]
+#[derive(Debug, PartialEq)]
```
Comment (eozturk1) [inline akd/src/errors.rs:14]: Why do we use `PartialEq` here?

File: akd_client/src/tests.rs [hunk 9]
```diff
@@ -255,7 +255,7 @@ async fn test_simple_lookup() -> Result<(), AkdError> {
         target_label_bytes,
         internal_lookup_proof,
     )
-    .map_err(|i_err| AkdError::AzksNotFound(format!("Internal: {:?}", i_err)));
+    .map_err(|i_err| AkdError::Storage(StorageError::Other(format!("Internal: {:?}", i_err))));
```
Comment (afterdusk) [inline akd_client/src/tests.rs:258]: What's the rationale for the Storage variant here?

File: akd/src/errors.rs [hunk 10]
```diff
@@ -45,9 +45,9 @@ impl From<DirectoryError> for AkdError {
     }
 }
 
-impl From<VRFStorageError> for AkdError {
-    fn from(error: VRFStorageError) -> Self {
-        Self::VRFStorageErr(error)
+impl From<VrfError> for AkdError {
```
Comment (eozturk1) [inline akd/src/errors.rs:48]: Do we lose any information on the error here? i.e., previously it was `VRF"Storage"Error` now it is only `VrfError`

File: akd/src/errors.rs [hunk 11]
```diff
@@ -57,15 +57,25 @@ impl From<AzksError> for AkdError {
...
+            AkdError::Directory(err) => {
+                writeln!(f, "AKD Directory Error: {}", err)
+            }
+            AkdError::AzksErr(err) => {
+                writeln!(f, "AKD AZKS Error: {}", err)
+            }
+            AkdError::Vrf(err) => {
+                writeln!(f, "AKD VRF Error: {}", err)
+            }
+            AkdError::Storage(err) => {
```
Comment (eozturk1) [inline akd/src/errors.rs:75]: Love this!

File: akd_mysql/src/mysql.rs [hunk 12]
```diff
@@ -794,7 +795,10 @@ impl Storage for AsyncMySqlDatabase {
 
         match self.internal_set(record, None).await {
             Ok(_) => Ok(()),
-            Err(error) => Err(StorageError::SetData(error.to_string())),
+            Err(error) => {
+                error!("MySQL error {}", error);
+                Err(StorageError::Other(format!("MySQL Error {}", error)))
```
Comment (afterdusk) [inline akd_mysql/src/mysql.rs:800]: nit: Maybe a `Database` or `Operation` variant? `Other` is fine too though!

File: akd/src/storage/memory.rs [hunk 13]
```diff
@@ -155,15 +155,19 @@ impl Storage for AsyncInMemoryDatabase {
                         return Ok(DbRecord::ValueState(item.clone()));
                     }
                 }
-                return Err(StorageError::GetData("Not found".to_string()));
+                return Err(StorageError::NotFound(format!("ValueState {:?}", id)));
```
Comment (eozturk1) [inline akd/src/storage/memory.rs:158]: nit: Maybe add "while getting `ValueState` directly from storage"?

File: akd/src/storage/memory.rs [hunk 14]
```diff
@@ -155,15 +155,19 @@ impl Storage for AsyncInMemoryDatabase {
...
             }
         }
         // fallback to regular get/set db
         let guard = self.db.read().await;
         if let Some(result) = (*guard).get(&bin_id).cloned() {
             Ok(result)
         } else {
-            Err(StorageError::GetData("Not found".to_string()))
+            Err(StorageError::NotFound(format!(
```
Comment (eozturk1) [inline akd/src/storage/memory.rs:166]: The origin of this error would be really helpful here when debugging. Can we clarify this is erroring out in `get_direct`?

File: akd/src/storage/memory.rs [hunk 15]
```diff
@@ -197,7 +201,7 @@ impl Storage for AsyncInMemoryDatabase {
 
             Ok(KeyData { states: results })
         } else {
-            Err(StorageError::GetData("Not found".to_string()))
+            Err(StorageError::NotFound(format!("ValueState {:?}", username)))
```
Comment (eozturk1) [inline akd/src/storage/memory.rs:204]: Same here. We should add the origins of the calls, i.e., getting user data.

File: akd/src/test_utils.rs [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+/// This is safe to call multiple times, but it will only initialize the logger
+/// to the log-level _first_ set. If you want a specific log-level (e.g. Debug)
+/// for a specific test, make sure to only run that single test after editing that
+/// test's log-level.
+///
+/// The default level applied everywhere is Info
+pub fn init_logger(level: Level) {
+    EPOCH.get_or_init(Instant::now);
+
+    INIT_ONCE.call_once(|| {
```
Comment (eozturk1) [inline akd/src/test_utils.rs:94]: Beautiful!

**Review by afterdusk — APPROVED** [review]
Great stuff as usual, thanks @slawlor ! Just some non-blocking nits and questions 🙂

File: akd/src/tests.rs [hunk 17]
```diff
@@ -713,3 +695,14 @@ async fn test_insert_single_leaf_full_tree() -> Result<(), HistoryTreeNodeError>
     assert!(root_val == expected, "Root hash not equal to expected");
     Ok(())
 }
+
+#[test]
+// Comment the following line to see log messages being written in failed testing scenarios
+#[should_panic(expected = "boom.")]
```
Comment (eozturk1) [inline akd/src/tests.rs:701]: Haha!

File: akd_mysql/src/mysql.rs [hunk 18]
```diff
@@ -794,7 +795,10 @@ impl Storage for AsyncMySqlDatabase {
...
+            }
```
Comment (eozturk1) [inline akd_mysql/src/mysql.rs:801]: Mention internal set?

**Review by eozturk1 — APPROVED** [review]
It looks great @slawlor! For some errors though I think we need more context (added some comments but there may be more). If you prefer to defer adding these contexts, I'd suggest keeping #157 open. I can work on it.

File: akd/src/storage/memory.rs (hunk 14 above)
Comment (slawlor) [reply akd/src/storage/memory.rs:166]: We could add log statements, but I specifically excluded "not found" results from log statements as it's a pretty common, expected flow. All of the storage errors which have real failures (i.e not "not found") include a log statement for the error.

File: akd/src/storage/memory.rs (hunk 13 above)
Comment (slawlor) [reply akd/src/storage/memory.rs:158]: See further down

File: akd_mysql/src/mysql.rs (hunk 12 above)
Comment (slawlor) [reply akd_mysql/src/mysql.rs:800]: Yeah `Other` is generally going to be some operation we could retry. If we needed (eventually) an error code which we _should not_ retry, then I'd say at that point to add something. But I think this is OK for now

**Pull request merged**